pkg/azureutils/logger.go

```diff
@@ -0,0 +1,189 @@
+package azureutils
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	consts "sigs.k8s.io/azuredisk-csi-driver/pkg/azureconstants"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2/klogr"
+	diskv1beta1 "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1beta1"
+)
+
+type contextKey int
+
+const workflowKey contextKey = 1
+
+type Workflow struct {
+	name         string
+	startTime    time.Time
+	requestID    string
+	logger       Logger
+	rootWorkflow *Workflow
+}
+
+func GetWorkflowFromObj(ctx context.Context, obj runtime.Object, details ...interface{}) (context.Context, Workflow) {
+	requestID := GetRequestID(obj)
+	startTimeStr := getAnnotationFromRuntimeObject(obj, consts.RequestStartimeKey)
+	var startTime time.Time
+	if startTimeStr != "" {
+		startTime, _ = time.Parse(consts.RequestTimeFormat, startTimeStr)
+	}
+	workflow := Workflow{
+		requestID: requestID,
+		startTime: startTime,
+		logger:    Logger{klogr.New().WithValues(append(GetObjectDetails(obj), consts.RequestIDKey, requestID)...)},
+	}
+	workflow.rootWorkflow = &workflow
+	ctx = workflow.SaveToContext(ctx)
+	return ctx, workflow
+}
+
+func GetObjectDetails(obj runtime.Object) []interface{} {
+	switch target := obj.(type) {
+	case *diskv1beta1.AzVolume:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.PvNameKey, target.Spec.PersistentVolume}
+		}
+	case *diskv1beta1.AzVolumeAttachment:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.NodeNameLabel, target.Spec.NodeName, consts.RoleLabel, target.Spec.RequestedRole}
+		}
+	}
+	return []interface{}{}
+}
+
+func StartWorkflow(ctx context.Context, name string, details ...interface{}) (context.Context, Workflow) {
+	w := Workflow{
+		requestID: GenerateRequestID(),
+	}
+	w.logger = Logger{klogr.New().WithValues(consts.RequestIDKey, w.requestID).WithValues(details...)}
+	w.rootWorkflow = &w
+	return w.StartChildWorkflow(ctx, name)
+}
+
+func (w Workflow) StartChildWorkflow(ctx context.Context, name string) (context.Context, Workflow) {
+	childWorkflow := Workflow{
+		rootWorkflow: w.rootWorkflow,
+		startTime:    time.Now(),
+		requestID:    w.requestID,
+		logger:       w.logger,
+		name:         name,
+	}
+	return childWorkflow.SaveToContext(ctx), childWorkflow
+}
+
+func emptyWorkflow() Workflow {
+	w := Workflow{
+		startTime: time.Now(),
+		logger:    Logger{klogr.New()},
+	}
+	w.rootWorkflow = &w
+	return w
+}
+
+func (w Workflow) Finish(err error) {
+	latency := time.Since(w.startTime)
+	w.logger.WithValues(consts.Latency, latency, "error", err, "workflow", w.name).WithCallDepth(1).Info("Workflow completed.")
```

Comment:
We should log error status through the `Error` function and omit a nil error in the success case.